Provide a custom Asciidoctor include macro

meistermeier opened the pull request:
uses a ruby based solution to add snippet blocks to the parent block element

Fixes gh-323

wilkinsona commented on spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java:
```diff
@@ -28,8 +30,15 @@
 	public void register(Asciidoctor asciidoctor) {
-		asciidoctor.javaExtensionRegistry()
-				.preprocessor(new DefaultAttributesPreprocessor());
+		final JavaExtensionRegistry registry = asciidoctor.javaExtensionRegistry();
```
REST Docs' code style is to only use `final` when it's required

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+#
+#   srd::operation-name[snippets='snippet-name1,snippet-name2', level=<indentation level>]
+#
+class RestDocsBlockMacro < Asciidoctor::Extensions::BlockMacroProcessor
```
I think this should be called `OperationBlockMacro` and the file itself should be called `operation_block_macro.rb`

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+    doc = parent.document
+    snippet_dir = doc.attributes['snippets']
+    snippets = snippets_from_attributes(attrs)
+    section_level = section_level_from_attributes(attrs)
```
I think we could use `parent.level` to figure this out automatically

wilkinsona commented on spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java:
```diff
@@ -28,8 +30,15 @@
+		rubyExtensionRegistry
+				.loadClass(RestDocsExtensionRegistry.class
+						.getResourceAsStream("/extensions/srd-include-block-macro.rb"))
+				.blockMacro("srd", "RestDocsBlockMacro");
```
I don't think the macro should be called `srd` as that would seem to imply the REST Docs will only ever have a single block macro. There doesn't seem to be a tradition of trying to namespace macro names in Asciidoctor so I'm tempted to go with `operation` instead.

wilkinsona commented on spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	@Test
+	public void useMacroWithoutSnippetAddsNothing() {
```
I'm intrigued by this test. What was the reasoning behind adding nothing? I would expect this to either produce an error or to include all of the snippets for a particular operation. My preference is for the latter but I'm not set on that.

meistermeier replied:
Now it adds every exiting snippet in the operation folder.

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+  end
+
+  def snippets_from_attributes(attrs)
+    (attrs['snippets'].split(',') if attrs['snippets']) || []
```
I think using [`fetch`](https://ruby-doc.org/core-1.9.3/Hash.html#method-i-fetch) would make this read a bit better:

```ruby
attrs.fetch('snippets', '').split(',')
```

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+
+  def write_content(content, snippet, params)
+    asciidoc_extension = '.adoc'
+    snippet_path = params[:snippet_dir].to_s + File::SEPARATOR +
```
I think using `File.join` and some string interpolation would improve the readability here. Something like:

```ruby
File.join(params[:snippet_dir], params[:operation], "#{snippet}.adoc")
```

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+    content.puts File.readlines(snippet_path).join
+
+  rescue Errno::ENOENT
+    content.puts "WARNING: snippet not found: " + snippet_path
```
I'd use string interpolation here too

wilkinsona commented on spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb:
```diff
@@ -0,0 +1,103 @@
+
+  def write_title(content, snippet, level)
+    section_level = '=' * level
+    content.puts section_level + ' ' + title(snippet)
```
I'd use string interpolation here too

wilkinsona commented on spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @author Gerrit Meier
+ */
+public class RestDocsIncludeBlockMacroTest {
```
This should be named `…Tests` to match the other tests classes

wilkinsona commented on spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java:
```diff
@@ -0,0 +1,165 @@
+
+	@Test
+	public void includeSnippetWithLevel() {
+		String expectedContent = "<div class=\"sect1\">\n" +
```
My preference would be to store the expected content in files in `src/test/resources`

meistermeier commented:
Thanks for the feedback. I think I can do the changes but this may take until the next weekend.

meistermeier commented:
I have updated the code and rebased it (so that the code comments above are broken sorry for that 🙂  ). 
Major changes:
* the include without any snippets will include every snippet in the operation directory
* the level of the include will be derived from the parent block level

wilkinsona commented:
Thanks, @meistermeier. Did you miss some files when you made the update? The new tests are failing for me because they can't find the `build/generated-snippets/some-operation` directory

meistermeier commented:
There went something completely wrong with the update...so I went down in the code dungeon and found some commits with loose endings ;)

wilkinsona commented:
@meistermeier I've merged this into master. Thank you very much for the contribution.

@mojavelinux We couldn't have done this without your help. Thank you very much.
